# Comparing Ask HN and Show HN posts from Hacker News
We're specifically interested in posts whose titles begin with either **Ask HN** or **Show HN**. Users submit Ask HN posts to ask the Hacker News community a specific question.
We'll compare these two types of posts to determine the following:

* Do Ask HN or Show HN receive more comments on average?
* Do posts created at a certain time receive more comments on average?

In [1]:
from csv import reader
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)
hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

## Removing Headers form the list of lists

In [2]:
headers = hn[0]
hn = hn[1:]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


## Extracting Ask HN and Show HN Posts

In [6]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(row)
    elif title.lower().startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


## Calculating the Average Number of Comments for Ask HN and Show HN Posts
Next, let's determine if ask posts or show posts receive more comments on average.

In [7]:
total_ask_comments = 0
for post in ask_posts:
    num_of_comments = int(post[4])
    total_ask_comments += num_of_comments
    
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)

14.038417431192661


In [9]:
total_show_comments = 0
for post in show_posts:
    num_of_comments = int(post[4])
    total_show_comments += num_of_comments
    
avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

10.31669535283993


On average, ask posts receive 14 comments while show posts recieve 10 comments. Ask posts recieve 4 more comments on average than show posts

Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

## Finding the Amount of Ask Posts and Comments by Hour Created
Next, we'll determine if we can maximize the amount of comments an ask post receives by creating it at a certain time. First, we'll find the amount of ask posts created during each hour of day, along with the number of comments those posts received. Then, we'll calculate the average amount of comments ask posts created at each hour of the day receive.

In [16]:
import datetime as dt

result_list = []

for post in ask_posts:
    result_list.append(
        [post[6], int(post[4])]
    )

comments_by_hour = {}
counts_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for each_row in result_list:
    date = each_row[0]
    comment = each_row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in counts_by_hour:
        comments_by_hour[time] += comment
        counts_by_hour[time] += 1
    else:
        comments_by_hour[time] = comment
        counts_by_hour[time] = 1

comments_by_hour

{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

## Calculating the Average Number of Comments for Ask HN Posts by Hour

In [17]:
avg_by_hour = []
for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])
avg_by_hour

[['14', 13.233644859813085],
 ['22', 6.746478873239437],
 ['16', 16.796296296296298],
 ['17', 11.46],
 ['06', 9.022727272727273],
 ['01', 11.383333333333333],
 ['21', 16.009174311926607],
 ['07', 7.852941176470588],
 ['05', 10.08695652173913],
 ['12', 9.41095890410959],
 ['11', 11.051724137931034],
 ['00', 8.127272727272727],
 ['10', 13.440677966101696],
 ['19', 10.8],
 ['20', 21.525],
 ['08', 10.25],
 ['15', 38.5948275862069],
 ['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['02', 23.810344827586206],
 ['23', 7.985294117647059],
 ['04', 7.170212765957447],
 ['03', 7.796296296296297],
 ['18', 13.20183486238532]]

## Sorting and Printing Values from a List of Lists

In [18]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])

print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse=True)
print("The top 5 for Ask Posts Comments")

for average, hour in sorted_swap:
    print("{}:{:.2f} average comments per post".format(
    dt.datetime.strptime(hr, "%H").strftime("%H:%M"), average))

[[13.233644859813085, '14'], [6.746478873239437, '22'], [16.796296296296298, '16'], [11.46, '17'], [9.022727272727273, '06'], [11.383333333333333, '01'], [16.009174311926607, '21'], [7.852941176470588, '07'], [10.08695652173913, '05'], [9.41095890410959, '12'], [11.051724137931034, '11'], [8.127272727272727, '00'], [13.440677966101696, '10'], [10.8, '19'], [21.525, '20'], [10.25, '08'], [38.5948275862069, '15'], [5.5777777777777775, '09'], [14.741176470588234, '13'], [23.810344827586206, '02'], [7.985294117647059, '23'], [7.170212765957447, '04'], [7.796296296296297, '03'], [13.20183486238532, '18']]
The top 5 for Ask Posts Comments
18:00:38.59 average comments per post
18:00:23.81 average comments per post
18:00:21.52 average comments per post
18:00:16.80 average comments per post
18:00:16.01 average comments per post
18:00:14.74 average comments per post
18:00:13.44 average comments per post
18:00:13.23 average comments per post
18:00:13.20 average comments per post
18:00:11.46 avera

The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.